// @kadima-tech/micro-service-base/lib/src/test-util/createTestApp.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createTestApp = void 0;
const fastify_1 = __importDefault(require("fastify"));
const cookie_1 = __importDefault(require("@fastify/cookie"));
const ajv_compiler_1 = __importDefault(require("@fastify/ajv-compiler"));
const fast_json_stringify_compiler_1 = __importDefault(require("@fastify/fast-json-stringify-compiler"));
const service_1 = require("../plugins/service");
const createTestApp = (opts) => {
    const ajvFactory = (0, ajv_compiler_1.default)();
    const app = (0, fastify_1.default)({
        ignoreTrailingSlash: true,
        schemaController: {
            compilersFactory: {
                buildValidator: ajvFactory,
                buildSerializer: (0, fast_json_stringify_compiler_1.default)(),
            },
        },
        ajv: {
            customOptions: {
                coerceTypes: "array",
                strictSchema: false,
            },
        },
        ...opts,
    });
    app.register(cookie_1.default);
    app.setErrorHandler(service_1.errorHandler);
    app.addHook("preHandler", service_1.injectContextHook);
    app.addHook("preSerialization", service_1.formatResponseHook);
    return app;
};
exports.createTestApp = createTestApp;
