// @kadima-tech/micro-service-base/lib/src/plugins/service/plugin.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fastify_1 = __importDefault(require("fastify"));
const plugin_1 = require("./plugin");
describe('Setup a service plugin', () => {
    it('Should return an async function to register with fastify', () => {
        const plugin = (0, plugin_1.setupServicePlugin)({
            routes: [],
        });
        expect(plugin).toBeDefined();
    });
    it('Should decorate the instance with a userContext', async () => {
        const route = {
            router: 'some-router',
            version: 'v1',
            path: '/some-path',
        };
        const plugin = (0, plugin_1.setupServicePlugin)({
            routes: [route],
        });
        const server = (0, fastify_1.default)();
        const registerSpy = jest.spyOn(server, 'register');
        await plugin(server, {});
        expect(registerSpy).toBeCalledWith(expect.anything(), {
            prefix: '/v1',
        });
    });
});
